File: Matlab/Rotinas_Ensaios/Ensaios.m
classdef Ensaios < handle
    
    properties
        mapeamento_experi %Tabela para relacionar experimento e numerico
    end
    
    methods (Abstract)
        [result,fail]=run(obj,mo_material,result)
        n=get_n_ensaio(obj) %Sendo fornecido no ensaio especifico
        sse=calc_sse(obj,experimento)
        time=get_time(obj)
        %fazer get t, para que nao seja necessario o ensaio ter exatamente
        %o objeto tempo. Desta forma deixo livre a forma que sera
        %forencido
        
    end
    
    methods
        
        function criar_mapeamento_experimentos(obj,time_experi)
            %Chamar
            %Verificacao do tempo de analise experimental e numerico
            time_numerico=obj.get_time();
            
            if max(time_experi)>max(time_numerico)%ficar atento max em matrizes
                error('Tempo de analise numerica menor que tempo experimental')
            end
            
            if min(time_experi)<min(time_numerico)%ficar atento max em matrizes
                error('Tempo de analise experimental menor que tempo numerio')
            end
            %% Cria vetor tempo corrigido, adicionando valores coincidentes com
            %% pontos experimentais
            %             ensaio.set_vetor_t_corrigido(time_experi)
%             for k=1:length(time_experi)
%                 for i=1:length(obj.t)-1
%                     if time_experi(k)>obj.t(i)&& time_experi(k)<obj.t(i+1)
%                         obj.t(i+1)= time_experi(k);
%                         obj.mapeamento_experi(k,1)=i+1;
%                         break
%                     end
%                 end
%             end
            % Metodo para criar tabela de mapemento entre valores experimentais
            % e valores numericos.
                        for i=1:length(time_experi)
                            for j=1:length( time_numerico)-1
                                if time_experi(i)>= time_numerico(j) && time_experi(i)<=time_numerico(j+1)
                                    obj.mapeamento_experi(i)=j;
                                    break %sai do loop menor, para nao continuar
                                end
                            end
                            %                 error('ponto nao encontrado')
                        end
            
        end %Fim metodo de mapeamento
    end %Fim metodos
end %Fim classe ensaios

File: ensaios.py
# -*- coding: utf-8 -*-
"""
Classe abstrata ensaios

@author: mathe
"""
from abc import ABC, abstractmethod
import sys
import numpy as np 

class Ensaios(ABC):
    
    @abstractmethod
    def run(self,material,result):
        """
        Metodo abstrato para rodar ensaio
        """
        pass
  #-------------------------------------------------------------         
  
    @abstractmethod
    def get_n_ensaio(self):
        """
        Metodo abstrato para fornecer n_steps do ensaio 
        """
        pass 
 #-------------------------------------------------------------         
   
    @abstractmethod
    def calc_sse(self,experimento):
        """
        Metodo abstrato para calculo do sse - Otimizacao
        """
        pass 
 #-------------------------------------------------------------         
   
    @abstractmethod
    def get_time(self):
        """
        Metodo abstrato para obter vetor tempo 
        """
        pass
#-------------------------------------------------------------         
   
    def criar_mapeamento_experimentos(self,time_experi):
        """
        Metodo para criar mapeamento entre experimento e numerico
        """
        time_numerico=self.get_time()
        max_time_experi=np.max(time_experi)
        max_time_numerico=np.max(time_numerico)
        
        min_time_experi=np.min(time_experi)
        min_time_numerico=np.min(time_numerico)

        # Verificacoes tamanho vetor numerico e experimental
        if (max_time_experi>max_time_numerico):
            sys.exit('Fatal Error: Tempo final de analise numerica menor que tempo final experimental')
            
        if (min_time_experi<min_time_numerico):
            sys.exit('Fatal Error: Tempo inicial de analise numerica menor que tempo inicial experimental')
        
        
        # Criando mapeamento entre valores experimentais e numericos 
        n_experi=len(time_experi)
        n_numeri=len(time_numerico)
        self.mapeamento_experi=np.zeros(n_experi)
        for i in range(n_experi):
            for j in range(n_numeri-1):
                if (time_experi[i]>=time_numerico[j]) and (time_experi[i]<=time_numerico[j+1]):
                    self.mapeamento_experi[i]=j
                    break 
#-------------------------------------------------------------         
        
        
            
            
        
        

